packages/ia-topnav/src/ia-topnav.js

```diff
@@ -0,0 +1,237 @@
+import './media-slider';
+import './desktop-subnav';
+import './dropdown-menu';
+import { user as userMenu } from './data/menus';
+import iaTopNavCSS from './styles/ia-topnav';
+
+export default class IATopNav extends LitElement {
+  static get styles() {
+    return iaTopNavCSS;
+  }
+
+  static get properties() {
+    return {
+      config: { type: Object },
+      mediaMenuAnimate: { type: Boolean },
+      mediaMenuOpen: { type: Boolean },
+      mediaSliderAnimate: { type: Boolean },
+      mediaSliderOpen: { type: Boolean },
+      searchMenuAnimate: { type: Boolean },
+      searchMenuOpen: { type: Boolean },
+      searchSubmitted: { type: Boolean },
+      selectedMenuOption: { type: String },
+      signedOutMenuOpen: { type: Boolean },
+      userMenuAnimate: { type: Boolean },
+      userMenuOpen: { type: Boolean },
+    };
+  }
+
+  constructor() {
+    super();
+    this.config = {};
+    this.mediaMenuAnimate = false;
+    this.mediaMenuOpen = false;
+    this.searchMenuAnimate = false;
+    this.searchMenuFade = false;
+    this.searchMenuOpen = false;
+    this.searchSubmitted = false;
+    this.signedOutMenuOpen = false;
+    this.userMenuAnimate = false;
+    this.userMenuOpen = false;
+    this.mediaSliderOpen = false;
+    this.mediaSliderAnimate = false;
+    this.selectedMenuOption = '';
+  }
+
+  updated(changedProperties) {
+    const { mediaMenuOpen, mediaSliderOpen } = this;
+    const menuClosed = changedProperties.has('mediaMenuOpen')
+      && changedProperties.get('mediaMenuOpen')
+      && !mediaMenuOpen;
+
+    if (menuClosed && mediaSliderOpen) {
+      this.mediaSliderOpen = false;
+      this.mediaSliderAnimate = false;
+      this.selectedMenuOption = '';
+    }
+  }
+
+  closeMediaSlider() {
+    this.mediaSliderAnimate = true;
+    this.mediaSliderOpen = false;
+    this.selectedMenuOption = '';
+  }
+
+  toggleMediaSlider() {
+    if (!this.mediaSliderOpen) {
+      this.mediaSliderAnimate = true;
+      this.mediaSliderOpen = !this.mediaSliderOpen;
+      this.mediaMenuOpen = true;
+    }
+  }
+
+  mediaMenu() {
+    this.userMenuOpen = false;
+    this.searchMenuOpen = false;
+    this.mediaMenuAnimate = true;
+    this.signedOutMenuOpen = false;
+    this.mediaMenuOpen = !this.mediaMenuOpen;
+  }
+
+  signedOutMenu() {
+    this.userMenuOpen = false;
+    this.searchMenuOpen = false;
+    this.mediaMenuOpen = false;
+    this.signedOutMenuOpen = !this.signedOutMenuOpen;
+  }
+
+  searchMenu() {
+    this.userMenuOpen = false;
+    this.mediaMenuOpen = false;
+    this.signedOutMenuOpen = false;
+    this.searchMenuAnimate = true;
+    this.searchMenuFade = true;
+    this.searchMenuOpen = !this.searchMenuOpen;
+    if (this.searchMenuOpen) {
+      window.setTimeout(() => {
+        this
+          .shadowRoot
+          .querySelector('primary-nav')
+          .shadowRoot
+          .querySelector('nav-search')
+          .shadowRoot
+          .querySelector('.search-field').focus();
```

Comment:
Similar comment about tunneling into components. If any of the child components change their component structure and change these, this will break.

As an alternative, you could call methods on child components to tell their children to give the search field focus. This allows components to only be responsible for their immediate children.
ie:
```
// ia-topnav.js
this.shadowRoot.querySelector('primary-nav').focusSearch();

// primary-nav.js
focusSearch() {
  this.shadowRoot.querySelector('nav-search').focusSearch();
}

// nav-search.js
focusSearch() {
  this.shadowRoot.querySelector('.search-field').focusSearch();
}
```

Reply:
Law of Demeter: "Only talk to your friends." (not to friends of friends)

Reply:
Your approach also breaks if the child components change. If, for example, the `<nav-search>` element is not rendered within `<primary-nav>`, a TypeError will be thrown.

I'll move it to a prop and pass it down through the components. This pattern is reminding me why I loved Redux so much.

Reply:
But if you remove `<nav-search>` from `<primary-nav>`, `<primary-nav>` is aware of that change since it's one of its own children and can be updated accordingly. It may change its behavior when `focusSearch()` is called, but `ia-topnav` doesn't need to be concerned about that.

It's about defining the API of each component and limiting your interaction to that.

Reply:
Switched to an `updated` method and a check for whether the component's `open`.